network/src/peers/peer/peer.rs

```diff
@@ -175,6 +190,78 @@ impl Peer {
     }
 
     pub(super) fn set_disconnected(&mut self) {
+        self.sync_state.reset();
         self.quality.disconnected();
     }
+
+    pub fn start_rtt_measurement(&mut self) {
+        self.quality.expecting_pong = true;
+        self.quality.last_ping_sent = Some(Instant::now());
+    }
+
+    pub fn stop_rtt_measurement(&mut self) {
+        if !self.quality.expecting_pong {
+            self.fail();
+
+            return;
+        }
+
+        let rtt = self
+            .quality
+            .last_ping_sent
+            .map(|x| x.elapsed().as_millis() as u64)
+            .unwrap_or(u64::MAX);
+
+        trace!("RTT for {} is {}ms", self.address, rtt);
+
+        self.quality.expecting_pong = false;
+        self.quality.rtt_ms = rtt;
+    }
+
+    pub fn register_received_message(&mut self) {
+        self.quality.see();
+        self.quality.num_messages_received += 1;
+    }
+
+    pub fn register_received_sync_block(&mut self) {
+        let remaining_sync_blocks = self.sync_state.remaining_sync_blocks;
+
+        // No need to cover the negative case since we saturate the decrement.
+        if remaining_sync_blocks == 0 {
+            trace!("received unexpected or late sync block from {}", self.address);
+
+            return;
+        }
+
+        // Decrement if > 0.
+        self.sync_state.remaining_sync_blocks = remaining_sync_blocks.saturating_sub(1);
```

Comment:
```suggestion
        self.sync_state.remaining_sync_blocks -= 1
```
this is fine, as we are already certain the value is greater than `0`